segregation/inference/randomization.py

```diff
@@ -0,0 +1,382 @@
+"""Tools for simulating spatial population distributions."""
+
+import itertools
+import multiprocessing
+from warnings import warn
+
+import geopandas as gpd
+import numpy as np
+import pandas as pd
+from joblib import Parallel, delayed
+from tqdm.auto import tqdm
+
+
+def _generate_estimate(input):
+    if hasattr(input[0], "_original_data"):
+        df = input[0]._original_data.copy()
+    else:
+        df = input[0].data.copy()
+    if input[0].index_type == "singlegroup":
+        df = input[1](df, group=input[0].group_pop_var, total=input[0].total_pop_var,)
+        estimate = (
+            input[0]
+            .__class__(df, input[0].group_pop_var, input[0].total_pop_var, **input[2])
+            .statistic
+        )
+    else:
+        df = input[1](df, groups=input[0].groups)
+        estimate = input[0].__class__(df, input[0].groups, **input[2]).statistic
+    return estimate
+
+
+def simulate_null(
+    iterations=500,
+    sim_func=None,
+    seg_class=None,
+    n_jobs=-1,
+    backend="loky",
+    index_kwargs=None,
+):
+    """Simulate a series of index values in parallel to serve as a null distribution.
+
+    Parameters
+    ----------
+    iterations : int, required
+        Number of iterations to simulate (size of the distribution), by default 1000
+    sim_func : function, required
+        population randomization function from segregation.inference to serve as
+        the null hypothesis.
+    seg_func : Class from segregation.singlegroup or segregation.singlegroup, required
+        fitted segregation class from which to generate a reference distribution
+    n_jobs : int, optional
+        number of cpus to initialize for parallelization. If -1, use all available,
+        by default -1
+    backend : str, optional
+        backend passed to joblib.Parallel, by default "loky"
+    index_kwargs : dict, optional
+        additional keyword arguments used to fit the index, such as distance or network
+        if estimating a spatial index; by default None
+
+    Returns
+    -------
+    list
+        pandas.Series of segregation indices for simulated data
+    """
+    if not index_kwargs:
+        index_kwargs = {}
+    if n_jobs == -1:
+        n_jobs = multiprocessing.cpu_count()
+    estimates = Parallel(n_jobs=n_jobs, backend=backend)(
+        delayed(_generate_estimate)((seg_class, sim_func, index_kwargs))
+        for i in tqdm(range(iterations))
```

Comment:
OH that's awesome, does tqdm work with joblib??

Reply:
I know i was surpised too. Saw on stack somewhere recently. It's not perfect, but good enough. Seems to fill the bar as each task is distributed to each worker, rather than when each task is completed, so the bar is like always 1 iteration ahead of the actual computation, but you kinda have to be looking for it to notice. Still gets the job done well enough